python/GafferUI/ColorChooser.py

```diff
@@ -157,6 +163,297 @@ def _displayTransformChanged( self ) :
+			staticComponent,
+			GafferUI.Slider.ValueChangedReason.DragMove
+		)
+		return True
+
+	def __dragEnd( self, widget, event ) :
+
+		c, staticComponent = self.getColor()
+		self.__setColorInternal(
+			self.__positionToColor( self.__clampPosition( event.line.p0 ) ),
+			staticComponent,
+			GafferUI.Slider.ValueChangedReason.DragEnd
+		)
+		return True
+
+	def __drawBackground( self, painter ) :
+
+		numStops = 50
+		if self.__colorFieldToDraw is None :
+			self.__colorFieldToDraw = QtGui.QImage( QtCore.QSize( numStops, numStops ), QtGui.QImage.Format.Format_RGB32 )
+
+			displayTransform = self.displayTransform()
+
+			xIndex, yIndex = self.__xyIndices()
+			zIndex = self.__zIndex()
+
+			staticValue = self.__color[zIndex]
+
+			c = imath.Color3f()
+			c[zIndex] = staticValue
+
+			ColorSpace = enum.Enum( "ColorSpace", [ "RGB", "HSV", "TMI" ] )
+			if self.__staticComponent in "rgb" :
+				colorSpace = ColorSpace.RGB
+			elif self.__staticComponent in "hsv" :
+				colorSpace = ColorSpace.HSV
+			else :
+				colorSpace = ColorSpace.TMI
+
+			xComponent, yComponent = self.__xyAxes()
+
+			for x in range( 0, numStops ) :
+				tx = float( x ) / ( numStops - 1 )
+				c[xIndex] = _ranges[xComponent].min + ( _ranges[xComponent].max - _ranges[xComponent].min ) * tx
+
+				for y in range( 0, numStops ) :
+					ty = float( y ) / ( numStops - 1 )
+
+					c[yIndex] = _ranges[yComponent].min + ( _ranges[yComponent].max - _ranges[yComponent].min ) * ty
+
+					if colorSpace == ColorSpace.RGB :
+						cRGB = c
+					elif colorSpace == ColorSpace.HSV :
+						cRGB = c.hsv2rgb()
+					else :
+						cRGB = _tmiToRGB( c )
+
+					cRGB = displayTransform( cRGB )
+					self.__colorFieldToDraw.setPixel( x, numStops - 1 - y, self._qtColor( cRGB ).rgb() )
+
+		painter.drawImage( self._qtWidget().contentsRect(), self.__colorFieldToDraw )
+
+	def __drawValue( self, painter ) :
+
+		position = self.__colorToPosition( self.__color )
+
+		pen = QtGui.QPen( QtGui.QColor( 0, 0, 0, 255 ) )
+		pen.setWidth( 1 )
+		painter.setPen( pen )
+
+		color = QtGui.QColor( 119, 156, 255, 255 )
+
+		painter.setBrush( QtGui.QBrush( color ) )
+
+		size = self.size()
+		if position.x >= 0 and position.y >= 0 and position.x <= size.x and position.y <= size.y :
+			painter.drawEllipse( QtCore.QPoint( position.x, position.y ), 4.5, 4.5 )
+			return
+
+		# Draw a triangle pointing out of the field. Prioritize corners to avoid cutting
+		# off part of the triangle out of the field when one axis is right at 0 or size.
+		#        |\  <-- We don't want this overhang.
+		# -------|->
+		#        |/|
+		#          |
+		#          |
+		if position.x >= size.x and position.y <= 0 :
+			rotation = -45.0  # Triangle pointing to the top-right
+		elif position.x <= 0 and position.y <= 0 :
+			rotation = -135.0  # Top-left
+		elif position.x <= 0 and position.y >= size.y :
+			rotation = -225.0  # Bottom-left
+		elif position.x >= size.x and position.y >= size.y :
+			rotation = -315.0  # Bottom-right
```

Comment:
I think the triangle code is a tiny bit over-zealous in some configurations, for instance in this one it suggests that B is out of range when it is actually just on the edge of the acceptable range :

![image](https://github.com/user-attachments/assets/57357351-e522-4264-b74f-288dbcf8278c)

( I would expect a left-facing triangle rather than a bottom-left facing triangle).

Reply:
In https://github.com/GafferHQ/gaffer/pull/5944/commits/2bddfe68bcbf6a9b01a44d8f059ab04f49e6755d I improved the readability of the out of range indicator. In cases like the one you pointed out, it uses a left facing triangle, added the same offset from the edge as the linear slider and kept the whole triangle in the field even when it's near enough to an edge that part of the triangle would otherwise get clipped off. A little less accurate but I think worth it for the legibility.